apps/condo/domains/banking/tasks/generateReports.js

```diff
@@ -0,0 +1,270 @@
+}
+
+async function updateTaskProgress ({ task, lastProgress, context, progress }) {
+    const taskId = get(task, 'id')
+    if (Date.now() - lastProgress > TASK_PROGRESS_UPDATE_INTERVAL) {
+        // User can cancel the task at any time, in this all operations should be stopped
+        task = await BankAccountReportTask.getOne(context, { id: taskId })
+        const taskStatus = get(task, 'status')
+        if (!task || taskStatus !== TASK_PROCESSING_STATUS) {
+            logger.info({ msg: 'status != processing. Aborting processing reports loop', taskStatus, taskSchemaName: BankAccountReportTask.gql.SINGULAR_FORM, taskId })
+            return
+        }
+
+        task = await BankAccountReportTask.update(context, taskId, {
+            ...DV_SENDER,
+            progress,
+        })
+    }
+}
+/**
+ * Aggregate transactions data and generate report
+ * @param taskId
+ * @returns {Promise<{categoryGroups: *[] }>}
+ */
+const generateReports = async (taskId) => {
+    if (!taskId) throw new Error('taskId is undefined')
+    const { keystone: context } = await getSchemaCtx('BankSyncTask')
+    let task = await BankAccountReportTask.getOne(context, { id: taskId })
+    if (!task) {
+        throw new Error(`Cannot find BankSyncTask by id="${taskId}"`)
+    }
+
+    const { account: { id: bankAccountId }, organization } = task
+
+    const bankAccount = await BankAccount.getOne(context, {
+        id: bankAccountId,
+        organization: { id: organization.id },
+        deletedAt: null,
+    })
+
+    const taskUpdatePayload = {
+        ...DV_SENDER,
+        progress: 0,
+    }
+    await BankAccountReportTask.update(context, taskId, taskUpdatePayload)
+
+    const allTransactions = await BankTransaction.getAll(context, {
+        deletedAt: null,
+        organization: { id: organization.id },
+        account: { id: bankAccount.id },
+    })
+
+    for (let transaction of allTransactions) {
+        if (!transaction.costItem) {
+            const contractorAccount = get(transaction, 'contractorAccount')
+            if (!contractorAccount){
+                throw new Error('contractorAccount is undefined')
+            }
+            transaction.costItem = { ...contractorAccount.costItem }
+        }
+    }
+
+    const sortedTransactions = sortTransactionsByMonthAndYear(allTransactions)
+    const monthTurnovers = calculateTransactionsTurnover(sortedTransactions)
+
+    let lastProgress = Date.now()
+    let costItemsDataSortedByDate = []
+    for (const transactions of sortedTransactions) {
+        const costItemData = {
+            date: dayjs(transactions[0].date).format('YYYY-MM'),
+            data: {},
+        }
+        for (let i = 0; i < transactions.length; i++) {
+            const currentTransaction = transactions[i]
+            if (Object.keys(costItemData.data).includes(currentTransaction.costItem.id)) {
+                costItemData.data[currentTransaction.costItem.id].amount += Number(currentTransaction.amount)
```

Comment:
Are we really want to have an object like `costItemData.data["some-uuid-value"].amount`?

Reply:
this is an intermediate format, used only in another internal utility.